barrels/LogMonkey/Logger.mc

```diff
@@ -0,0 +1,172 @@
+
+module LogMonkey {
+
+    //! The Logger class provides public APIs to log various types of
+    //! messages. Each Logger class is associated with a log level which
+    //! is set within the Logger.initialize() function.
+    class Logger {
+
+        private var mLogLevel, mLogStream;
+
+        //! Creates a new Logger object.
+        //! @param [Toybox::Lang::String] The log level value to log messages through this class to
+        //! @param [Toybox::Lang::Object] An object which defines a println(message) function
+        (:debug)
+        function initialize(logLevel, logStream) {
+            mLogLevel = logLevel;
+            mLogStream = logStream;
+        }
+
+        // An empty version of all functions in this class annotated
+        // with the :release tag are included to save memory when an
+        // app is compiled in release mode. We will still have the same
+        // symbols available in the release build so we won't get symbol
+        // not found errors while saving memory.
+        (:release)
+        function initialize(logLevel, logStream) {
+        }
+
+        // The string formats to use when printing log messages
+        private static var FORMAT_VARIABLE = "$1$=($2$) $3$";
+        private static var FORMAT_LOG_MESSAGE = "(lmf1)[$1$] {$2$} $3$: $4$";
+        private static var FORMAT_TIMESTAMP = "$1$-$2$-$3$ $4$:$5$:$6$";
+
+        //! Forms a log message based on the given values.
+        //! @param [Toybox::Lang::String] tag The tag to apply to the log message
+        //! @param [Toybox::Lang::String] message The message to log
+        //! @return [Toybox::Lang::String] A log message matching the format specified by FORMAT_TIMESTAMP
+        (:debug)
+        private function formLogMessage(tag, message) {
+            // Get a timestamp from the system
+            var currentTime = Greg.info(Time.now(), Time.FORMAT_SHORT);
+            var timestamp = Lang.format(FORMAT_TIMESTAMP, [
+                currentTime.month.format("%02u"),
+                currentTime.day.format("%02u"),
+                currentTime.year.format("%04u"),
+                currentTime.hour.format("%02u"),
+                currentTime.min.format("%02u"),
+                currentTime.sec.format("%02u")
+            ]);
+
+            // Form the log message
+            return Lang.format(FORMAT_LOG_MESSAGE, [timestamp, mLogLevel, tag, message]);
+        }
+
+        (:release)
+        private function formLogMessage(tag, message) {
+        }
+
+        //! Handles the printing of a log message to the log file
+        //! @param [Toybox::Lang::String] tag The tag to apply to the log message
+        //! @param [Toybox::Lang::String] message The message to log
+        (:debug)
+        private function log(tag, message) {
+            // Print the log message
+            mLogStream.println(formLogMessage(tag, message));
+        }
+
+        (:release)
+        private function log(tag, message) {
```

Comment:
This method is not needed in release. It should probably be omitted.

Reply:
Done